docs/architecture/ADR-001-ABCI++.md

```diff
@@ -0,0 +1,413 @@
+   }
+
+   sort.Slice(messages.MessagesList, func(i, j int) bool {
+       return bytes.Compare(messages.MessagesList[i].NamespaceId, messages.MessagesList[j].NamespaceId) < 0
+   })
+
+   return sqwr.export(), &core.Data{
+       Txs:                    processedTxs,
+       Messages:               messages,
+       Evidence:               data.Evidence,
+       IntermediateStateRoots: data.IntermediateStateRoots,
+   }, nil
+}
+
+...
+
+// writeTx marshals the tx and lazily writes it to the square. Returns true if
+// the write was successful, false if there was not enough room in the square.
+func (sqwr *squareWriter) writeTx(tx []byte) (ok bool, err error) {
+   delimTx, err := coretypes.Tx(tx).MarshalDelimited()
+   if err != nil {
+       return false, err
+   }
+
+   if !sqwr.hasRoomForTx(delimTx) {
+       return false, nil
+   }
+
+   sqwr.txWriter.Write(delimTx)
+   return true, nil
+}
+
+// writeMalleatedTx malleates a MsgWirePayForMessage into a MsgPayForMessage and
+// its corresponding message provided that it has a MsgPayForMessage for the
+// preselected square size. Returns true if the write was successful, false if
+// there was not enough room in the square.
+func (sqwr *squareWriter) writeMalleatedTx(
+   parentHash []byte,
+   tx signing.Tx,
+   wpfm *types.MsgWirePayForMessage,
+) (ok bool, malleatedTx coretypes.Tx, msg *core.Message, err error) {
+   ... // process the malleated tx and extract the message.
+
+   // check if we have room for both the tx and message it is crucial that we
+   // add both atomically, otherwise the block is invalid
+   if !sqwr.hasRoomForBoth(wrappedTx, coreMsg.Data) {
+       return false, nil, nil, nil
+   }
+
+   ...
+
+   sqwr.txWriter.Write(delimTx)
+   sqwr.msgWriter.Write(coretypes.Message{
+       NamespaceID: coreMsg.NamespaceId,
+       Data:        coreMsg.Data,
+   })
+
+   return true, wrappedTx, coreMsg, nil
+}
+```
+
+Lastly, the data availability header is used to create the `DataHash` in the `Header` in the application instead of in tendermint. This is done by modifying the protobuf version of the block data to retain the cached hash and setting it during `ProcessProposal`. Later, in `ProcessProposal` other fullnodes check that the `DataHash` matches the block data by recomputing it. Previously, this extra check was performed inside the `ValidateBasic` method of `types.Data`, where is was computed each time it was decoded. Not only is this more efficient as it saves significant computational resources and keeps `ValidateBasic` light, it is also much more explicit. This approach does not however dramatically any existing code in tendermint, as the code to compute the hash of the block data remains there. Ideally, we would move all of the code that computes erasure encoding to the app. This approach allows us to keep the intuitiveness of the `Hash` method for `types.Data`, along with not forcing us to change many tests in tendermint, which rely on this functionality.
```

Comment:
This approach does not however dramatically **change** any existing code in tendermint
```suggestion
Lastly, the data availability header is used to create the `DataHash` in the `Header` in the application instead of in tendermint. This is done by modifying the protobuf version of the block data to retain the cached hash and setting it during `ProcessProposal`. Later, in `ProcessProposal` other fullnodes check that the `DataHash` matches the block data by recomputing it. Previously, this extra check was performed inside the `ValidateBasic` method of `types.Data`, where is was computed each time it was decoded. Not only is this more efficient as it saves significant computational resources and keeps `ValidateBasic` light, it is also much more explicit. This approach does not however dramatically change any existing code in tendermint, as the code to compute the hash of the block data remains there. Ideally, we would move all of the code that computes erasure encoding to the app. This approach allows us to keep the intuitiveness of the `Hash` method for `types.Data`, along with not forcing us to change many tests in tendermint, which rely on this functionality.
```